src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -81,42 +92,58 @@
    watcher :- (schema/protocol Watcher)]
   (let [dir-create? (fn [event]
                       (and (= :create (:type event))
-                           (fs/directory? (:path event))))]
+                           (fs/directory? (:full-path event))))]
     (DirWatchUtils/registerRecursive (:watch-service watcher)
                                      (->> events
                                           (filter dir-create?)
-                                          (map #(.toPath (:path %)))))))
+                                          (map #(.toPath (:full-path %)))))))
+
+(schema/defn get-event-maps-from-key :- [Event]
+  [watch-key :- WatchKey]
+  (let [events (.pollEvents watch-key)]
+    (map #(clojurize % (.watchable watch-key)) events)))
 
-(schema/defn retrieve-events
-  :- [(schema/one WatchKey "key") (schema/one [WatchEvent] "events")]
-  "Blocks until an event the watcher is concerned with has occured.
-  Returns the native WatchKey and WatchEvents"
+(schema/defn retrieve-events :- [Event]
+  "Blocks until an event the watcher is concerned with has occured. Will then
+  poll for a new event, waiting at least `window-min` for a new event to
+  occur. Will continue polling for as long as there are new events that occur
+  within `window-min`, or the `window-max` time limit has been exceeded."
   [watcher :- (schema/protocol Watcher)]
   (let [watch-key (.take (:watch-service watcher))
-        events (.pollEvents watch-key)]
+        events (get-event-maps-from-key watch-key)
+        time-limit (+ (System/currentTimeMillis) window-max)]
     (.reset watch-key)
-    [watch-key events]))
+    (watch-new-directories! events watcher)
+      (loop [events' events]
+        (if-let [waiting-key (.poll (:watch-service watcher) window-min window-units)]
+          (let [waiting-events (get-event-maps-from-key waiting-key)]
+            (watch-new-directories! waiting-events watcher)
+            (.reset waiting-key)
```

Comment:
Good question! My understanding is:

We want to call `watch-new-directories` asap so that there's as little chance as possible that we're missing nested directory creation.

We want to call `reset` before we attempt to `take` or `poll` for a `watch-key` so we can potentially get new events from it.

We don't need to call `watch-new-directories` if `events` is empty, though it shouldn't hurt.
We need to call `reset` even if `events` is empty.

I think I should switch the order above this loop.